internal/sql/rewrite/parameters.go

```diff
@@ -41,55 +41,61 @@ func isNamedParamSignCast(node ast.Node) bool {
 	return astutils.Join(expr.Name, ".") == "@" && cast
 }
 
-func NamedParameters(engine config.Engine, raw *ast.RawStmt, numbs map[int]bool, dollar bool) (*ast.RawStmt, map[int]string, []source.Edit) {
+// paramFromName takes a user-defined parameter name, with an optional suffix of
+// ? (nullable), or ! (non-null) and builds the appropiate parameter
+func paramFromName(name string) named.Param {
+	if len(name) == 0 {
+		return named.NewUnspecifiedParam(name)
+	}
+
+	last := name[len(name)-1]
+	if last == '!' {
+		return named.NewUserDefinedParam(name[:len(name)-1], true)
+	}
+
+	if last == '?' {
+		return named.NewUserDefinedParam(name[:len(name)-1], false)
+	}
+
+	return named.NewUnspecifiedParam(name)
+}
+
+func NamedParameters(engine config.Engine, raw *ast.RawStmt, numbs map[int]bool, dollar bool) (*ast.RawStmt, *named.ParamSet, []source.Edit) {
 	foundFunc := astutils.Search(raw, named.IsParamFunc)
 	foundSign := astutils.Search(raw, named.IsParamSign)
+	hasNamedParameterSupport := engine != config.EngineMySQL
+	allParams := named.NewParamSet(numbs, hasNamedParameterSupport)
+
 	if len(foundFunc.Items)+len(foundSign.Items) == 0 {
-		return raw, map[int]string{}, nil
+		return raw, allParams, nil
 	}
 
-	hasNamedParameterSupport := engine != config.EngineMySQL
-
-	args := map[string][]int{}
-	argn := 0
 	var edits []source.Edit
 	node := astutils.Apply(raw, func(cr *astutils.Cursor) bool {
 		node := cr.Node()
 		switch {
 		case named.IsParamFunc(node):
 			fun := node.(*ast.FuncCall)
-			param, isConst := flatten(fun.Args)
-			if nums, ok := args[param]; ok && hasNamedParameterSupport {
-				cr.Replace(&ast.ParamRef{
-					Number:   nums[0],
-					Location: fun.Location,
-				})
-			} else {
-				argn++
-				for numbs[argn] {
-					argn++
-				}
-				if _, found := args[param]; !found {
-					args[param] = []int{argn}
-				} else {
-					args[param] = append(args[param], argn)
-				}
-				cr.Replace(&ast.ParamRef{
-					Number:   argn,
-					Location: fun.Location,
-				})
-			}
```

Comment:
_aside_: encapsulating the positional tracking into `named.ParamSet` significantly cleaned up this code.